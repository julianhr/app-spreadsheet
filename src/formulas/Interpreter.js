import { Lexer } from './Lexer'
import { parseLocation } from '~/library/utils'
import graph from '~/formulas/graph'
import Parser from './Parser'
import ReduxConnect from './ReduxConnect'


const ERR_DIVISION_BY_ZERO = '#DIV/0!'
const ERR_CIRCULAR_REFERENCE = '#REF!'
const ERR_GENERIC = '#ERROR!'

class Interpreter {
  constructor(location) {
    this.location = location
    this.error = null
    this.cache = {}
    this.result = null
    this.state = new ReduxConnect()
  }

  interpret(input) {
    const vertex = graph.addVertex(this.location, input)
    let lexer, parser

    try {
      lexer = new Lexer(input)
      parser = new Parser( lexer.getTokens() )
      vertex.ast = parser.parse()
      graph.recalculate(this.location)
      this.result = vertex.result
    } catch (error) {
      this.error = vertex.error = error
      this.result = vertex.result = this.errorResponse()
      return this.result
    }

    this.error = vertex.error
    return this.result
  }

  visit(location) {
    const vertex = graph.getVertex(location)

    if (graph.isResolved(location)) {
      return vertex.result
    }

    graph.markVisited(location)

    if (!vertex.ast) {
      return vertex.result
    }

    try {
      this.result = vertex.result = this._visit(vertex.ast)
    } catch (error) {
      this.error = vertex.error = error
      this.result = vertex.result = this.errorResponse()
      throw error
    }

    return this.result
  }

  _visit(node) {
    switch (node._name) {
      case 'NumberNode':
        return this.NumberNode(node)
      case 'CellNode':
        return this.CellNode(node)
      case 'TextNode':
        return this.TextNode(node)
      case 'FuncOp':
        return this.FuncOp(node)
      case 'BinaryOp':
        return this.BinaryOp(node)
      case 'UnaryOp':
        return this.UnaryOp(node)
      default:
        const nodeName = node._name || (node.constructor || {}).name
        throw new Error(`Unrecognized AST node ${nodeName}`)
    }
  }

  TextNode(node) {
    return node.eval()
  }

  NumberNode(node) {
    return node.eval()
  }

  CellNode(node) {
    const otherLocation = node.location
    let result

    if (!this.isLocationValid(otherLocation)) {
      throw new Error('Location out of bounds')
    }

    if (this.cache[otherLocation] !== undefined) {
      return this.cache[otherLocation]
    }

    if (graph.isVisited(otherLocation)) {
      throw new Error('Circular reference')
    }

    result = this.dfsCellVisit(otherLocation)
    this.cache[otherLocation] = result
    return result
  }

  UnaryOp(node) {
    const result = this._visit(node.expr)
    return node.eval(result)
  }

  FuncOp(node) {
    const evaluatedArgs = []

    for (let childNode of node.argNodes) {
      const result = this._visit(childNode)
      evaluatedArgs.push(result)
    }

    return node.eval(evaluatedArgs)
  }

  BinaryOp(node) {
    const left = this._visit(node.leftNode)
    const right = this._visit(node.rightNode)
    return node.eval(left, right)
  }

  isLocationValid(location) {
    const [colIndex, rowIndex] = parseLocation(location)

    if (Math.min(colIndex, rowIndex) < 0) { return false }
    if (colIndex >= this.state.columns) { return false }
    if (rowIndex >= this.state.rows) { return false }
    return true
  }

  dfsCellVisit(otherLocation) {
    const interpreter = new Interpreter(otherLocation)
    let result = 0

    // cell has a value
    if (graph.hasVertex(otherLocation)) {
      result = interpreter.visit(otherLocation)
    }

    return result
  }

  errorResponse() {
    const { message: errorMsg } = this.error

    if (errorMsg.match(/division by zero/i)) {
      return ERR_DIVISION_BY_ZERO
    } else if (errorMsg.match(/circular reference/i)) {
      return ERR_CIRCULAR_REFERENCE
    } else {
      return ERR_GENERIC
    }
  }
}

export default Interpreter
export {
  ERR_DIVISION_BY_ZERO,
  ERR_CIRCULAR_REFERENCE,
  ERR_GENERIC,
}
