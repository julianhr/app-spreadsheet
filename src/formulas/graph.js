import ReduxConnect from './ReduxConnect'
import Interpreter from './Interpreter'


class Vertex {
  constructor(entered) {
    this.entered = entered
    this.error = null
    this.result = null
    this.ast = null
  }
}

class Graph {
  constructor() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  interpret(location, input) {
    const interpreter = new Interpreter(location)
    return interpreter.interpret(input)
  }

  recalculate(initLocation) {
    this.setPendingNodes()
    this.dfs(initLocation)
    this.updateStore()
  }

  getCellResult(location) {
    if (this.hasVertex(location)) {
      return this.getVertex(location).result
    } else {
      return ''
    }
  }

  setPendingNodes() {
    this.pending = new Set()

    Object.keys(this.adj).forEach(vertex => {
      this.pending.add(vertex)
    })
  }

  isResolved(location) {
    return this.isVisited(location) || !this.isPending(location)
  }

  isVisited(location) {
    return this.visited.has(location)
  }

  isPending(location) {
    return this.pending.has(location)
  }

  dfs(location) {
    while (this.pending.size > 0) {
      let interpreter
      this.visited = new Set()

      if (!location) {
        location = [...this.pending][0]
      }

      try {
        interpreter = new Interpreter(location)
        interpreter.visit(location)
      } catch (error) {}

      this.visited.forEach(location => this.pending.delete(location))
      location = null
    }
  }

  updateStore() {
    const locationKeys = new Set(this.store.locations)

    for (let [location, vertex] of Object.entries(this.adj)) {
      const { entered, result } = vertex

      locationKeys.delete(location)

      if (result !== this.store.getCellResult(location)) {
        this.store.replaceCellData(location, entered, result)
      }
    }

    for (let location of locationKeys) {
      this.store.clearCell(location)
    }
  }

  markVisited(location) {
    return this.visited.add(location)
  }

  addVertex(location, entered) {
    const vertex = new Vertex(entered)
    this.adj[location] = vertex
    return vertex
  }

  hasVertex(location) {
    return location in this.adj
  }

  getVertex(location) {
    this.testMissingLocation(location)
    return this.adj[location]
  }

  delVertex(location) {
    if (location in this.adj) {
      delete this.adj[location]
      this.recalculate()
    }
  }

  resetAll() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  testMissingLocation(location) {
    if (!this.hasVertex(location)) {
      throw new Error('Missing location')
    }
  }
}

const graph = new Graph()


export default graph
export {
  Vertex,
  Graph,
}
