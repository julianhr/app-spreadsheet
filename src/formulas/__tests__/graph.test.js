import graph from '../graph'
import Interpreter from '../Interpreter'


describe('graph', () => {
  beforeEach(() => {
    graph.resetAll()
  })

  test('#interpret ', () => {
    jest.spyOn(Interpreter.prototype, 'interpret')
    const result = graph.interpret('A-1', '=5+2')
    expect(result).toBe(7)
    expect(Interpreter.prototype.interpret).toHaveBeenCalledTimes(1)
  })

  test('#recalculate', () => {
    graph.interpret('A-1', '=5+2')
    jest.spyOn(graph, 'setPendingNodes')
    jest.spyOn(graph, 'dfs')
    jest.spyOn(graph, 'updateStore')

    graph.recalculate('A-1')
    expect(graph.setPendingNodes).toHaveBeenCalledTimes(1)
    expect(graph.dfs).toHaveBeenCalledTimes(1)
    expect(graph.updateStore).toHaveBeenCalledTimes(1)
  })

  describe('#getCellResult', () => {
    it('returns value if vertex exists', () => {
      graph.interpret('A-1', '=5+2')
      expect(graph.getCellResult('A-1')).toBe(7)
    })

    it('returns empty string if vertex does not exist', () => {
      expect(graph.getCellResult('A-1')).toBe('')
    })
  })

  test('#setPendingNodes', () => {
    graph.interpret('A-1', '=5+2')
    graph.interpret('A-2', '=A1+3')
    graph.interpret('B-2', '5')
    graph.setPendingNodes()
    expect(graph.pending).toEqual(new Set(['A-1', 'A-2', 'B-2']))
  })

  describe('#isResolved', () => {
    it('identifies if vertex was already processed during graph recalculation', () => {
      graph.interpret('A-1', '5')
      graph.interpret('A-2', '=A1')
      graph.interpret('A-3', '=A1')
      graph.interpret('B-3', '=A1')
      jest.spyOn(graph, 'isResolved')

      graph.recalculate('A-2')
      const mockResults = []
      graph.isResolved.mock.calls.forEach((call, i) =>
        mockResults.push([call[0], graph.isResolved.mock.results[i]]))
      expect(mockResults).toMatchSnapshot()
    })
  })
})
