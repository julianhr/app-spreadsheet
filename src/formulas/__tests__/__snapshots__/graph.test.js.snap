// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph #isResolved identifies if vertex was already processed during graph recalculation 1`] = `
Array [
  Array [
    "A-2",
    Object {
      "type": "return",
      "value": false,
    },
  ],
  Array [
    "A-1",
    Object {
      "type": "return",
      "value": false,
    },
  ],
  Array [
    "A-3",
    Object {
      "type": "return",
      "value": false,
    },
  ],
  Array [
    "A-1",
    Object {
      "type": "return",
      "value": true,
    },
  ],
  Array [
    "B-3",
    Object {
      "type": "return",
      "value": false,
    },
  ],
  Array [
    "A-1",
    Object {
      "type": "return",
      "value": true,
    },
  ],
]
`;
